CONTRIBUTING.md

```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
+messages must still provide enough detail to diagnose the sorts of problems
+that Elasticsearch might encounter. In some cases it works well to collect
+information over a period of time and then log a complete summary, rather than
+recording every step of a process in its own message.
+
+It's possible that the reader of `DEBUG` logs is also reading the code, but
+that is less likely than for `TRACE` logs. You should strive to avoid
+terminology that only makes sense when reading the code, and also aim for
+messages at this level to be self-contained rather than intending them to be
+read as a sequence.
+
+It's often useful to log exceptions and other deviations from the "happy path"
+at `DEBUG` level. Exceptions logged at `DEBUG` should generally include the
+complete stack trace.
+
+##### `INFO`
+
+This is the next least verbose level, and the first level that is enabled by
+default. It is appropriate for recording important events in the life of the
+cluster, such as an index being created or deleted, or a snapshot starting or
+completing. Users will mostly ignore log messages at `INFO` level, but may use
+these messages to construct a high-level timeline of events leading up to an
+incident.
+
+This level is enabled by default so its target audience is the general
+population of users and administrators. You should use user-facing terminology
+and ensure that messages at this level are self-contained. In general you
+shouldn't log unusual events, particularly exceptions with stack traces, at
+`INFO` level. If the event is relatively benign then use `DEBUG`, whereas if
+the user should be notified then use `WARN`.
+
+##### `WARN`
+
+This is the next least verbose level, and is also enabled by default. Ideally a
+healthy cluster will emit no `WARN`-level logs, but this is the appropriate
+level for recording events that the cluster administrator should investigate,
+or which indicate a bug. Some production environments require the cluster to
+emit no `WARN`-level logs during acceptance testing, so you must ensure that
+any logs at this level really do indicate a problem that needs addressing.
+
+As with the `INFO` level, you should use user-facing terminology at the `WARN`
+level, and also ensure that messages are self-contained. Strive to make them
+actionable too since you should be logging at this level when the user should
+take some investigative action. Unlike at the `INFO` level, it is often
+appropriate to log an exception, complete with stack trace, at `WARN` level.
+Although the stack trace may not be useful to the user, it may contain
+information that is vital for a developer to fully understand the problem. In
+other cases it may be appropriate to log the exception message at `WARN` and
+only log the full exception if the user has enabled `DEBUG` logging:
+
+    if (logger.isDebugEnabled()) {
+        logger.warn("investigate me", exception);
+    } else {
+        logger.warn("investigate me: [{}]", exception.getMessage())
+    }
```

Comment:
Do we only do this when we think that the stack trace is not generally useful? Because the cause is super well understood, I guess. It's kind of bothered me that you have to read the code to know that setting the thing to debug will cause the whole stack trace to be logged.

Anyway! I think that this pattern is a bit sneaky and I'd love to see examples of where we use it.

Reply:
Sure, the most common place I use this is where the exception is almost always some boring network problem, like here:

https://github.com/elastic/elasticsearch/blob/68817d7ca29c264b3ea3f766737d81e2ebb4028c/server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java#L172-L176

Logging the stack trace here doesn't help, but it does generate a lot of noise when there's a network partition and that leads folks to report _just this message_ when asking for help.

On reflection maybe we should make this decision by looking at the exception too, only suppressing the stack trace if it really is a boring network problem.

Reply:
The other case is where the error is reported to the client, but we want to record it in the logs as well.

That's rare (we don't usually log client errors), but if it's an administrative action that's failed, sometimes we want the logs to mention it, but having a full stack trace each time isn't applicable, except for debug purposes.

Reply:
In 6942b041d2a13741434689ca599f49e811ceb858 I adjusted the guidance here to only suppress stack traces on boring exceptions. I opened https://github.com/elastic/elasticsearch/pull/75671 since the stack trace on `ReceiveTimeoutTransportException` is always boring anyway, and I'll align the code I quoted above with this new guidance at some point.

I also added more nuance in e1935754672401b360fed1328286b657dd3427b2 to the guidance about logging exceptions that directly result from client requests.

Reply:
Thanks!